cairo/src/utils/signature.cairo

```diff
@@ -1,44 +1,364 @@
+from src.utils.circuit_utils import (
+    N_LIMBS,
+    hash_full_transcript_and_get_Z_3_LIMBS,
+    scalar_to_epns,
+    felt_to_UInt384,
+    run_modulo_circuit_basic,
+)
+
+from src.utils.ecdsa_circuit import (
+    get_full_ecip_2P_circuit,
+    get_ADD_EC_POINT_circuit,
+    get_DOUBLE_EC_POINT_circuit,
+)
+
+struct G1Point {
+    x: UInt384,
+    y: UInt384,
+}
+
+namespace secp256k1 {
+    const CURVE_ID = 2;
+    const P0 = 0xfffffffffffffffefffffc2f;
+    const P1 = 0xffffffffffffffffffffffff;
+    const P2 = 0xffffffffffffffff;
+    const P3 = 0x0;
+    const N0 = 0xaf48a03bbfd25e8cd0364141;
+    const N1 = 0xfffffffffffffffebaaedce6;
+    const N2 = 0xffffffffffffffff;
+    const N_LOW_128 = 0xbaaedce6af48a03bbfd25e8cd0364141;
+    const N_HIGH_128 = 0xfffffffffffffffffffffffffffffffe;
+    const N3 = 0x0;
+    const A0 = 0x0;
+    const A1 = 0x0;
+    const A2 = 0x0;
+    const A3 = 0x0;
+    const B0 = 0x7;
+    const B1 = 0x0;
+    const B2 = 0x0;
+    const B3 = 0x0;
+    const G0 = 0x3;
+    const G1 = 0x0;
+    const G2 = 0x0;
+    const G3 = 0x0;
+    const MIN_ONE_D0 = 0xfffffffffffffffefffffc2e;
+    const MIN_ONE_D1 = 0xffffffffffffffffffffffff;
+    const MIN_ONE_D2 = 0xffffffffffffffff;
+    const MIN_ONE_D3 = 0x0;
+}
+
+const POW_2_32 = 2 ** 32;
+const POW_2_64 = 2 ** 64;
+const POW_2_96 = 2 ** 96;
+
+@known_ap_change
+func get_generator_point() -> (point: G1Point) {
+    // generator_point = (
+    //     0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798,
+    //     0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8
+    // ).
+    return (
+        point=G1Point(
+            x=UInt384(
+                0x2dce28d959f2815b16f81798, 0x55a06295ce870b07029bfcdb, 0x79be667ef9dcbbac, 0x0
+            ),
+            y=UInt384(
+                0xa68554199c47d08ffb10d4b8, 0x5da4fbfc0e1108a8fd17b448, 0x483ada7726a3c465, 0x0
+            ),
+        ),
+    );
+}
+
+@known_ap_change
+func sign_to_UInt384_mod_secp256k1(sign: felt) -> (res: UInt384) {
+    if (sign == -1) {
+        return (res=UInt384(secp256k1.MIN_ONE_D0, secp256k1.MIN_ONE_D1, secp256k1.MIN_ONE_D2, 0));
+    } else {
+        return (res=UInt384(1, 0, 0, 0));
+    }
+}
+
+// Input must be a valid Uint256.
+func uint256_to_uint384{range_check_ptr}(a: Uint256) -> (res: UInt384) {
+    let (high_64_high, high_64_low) = divmod(a.high, POW_2_64);
+    let (low_32_high, low_96_low) = divmod(a.low, POW_2_96);
+    return (res=UInt384(low_96_low, low_32_high + POW_2_32 * high_64_low, high_64_high, 0));
+}
+
+// Assume the input is valid UInt384 (will be the case if coming from ModuloBuiltin)
+func uint384_to_uint256_mod_p{range_check_ptr}(a: UInt384, p: UInt384) -> (res: Uint256) {
+    // First force the prover to have filled a fully reduced field element < P.
+    assert a.d3 = 0;
+    assert [range_check_ptr] = p.d2 - a.d2;  // a.d2 <= p.d2
+    tempvar range_check_ptr = range_check_ptr + 1;
```

Comment:
```suggestion
    // Assert that a <= P.
    assert a.d3 = 0;
    assert p.d3 = 0;
    assert [range_check_ptr] = p.d2 - a.d2;  // a.d2 <= p.d2
    tempvar range_check_ptr = range_check_ptr + 1;
```

I really discourage the usage of "the prover xxx" as it's most of the time misleading. The program should speak for itself.

Reply:
this could actually go in the `else` 

```
assert a.d3 = p.d3; // actually 0
if (a.d2 == p.d2) {
    if (a.d1 == p.d1) {
        assert [range_check_ptr] = p.d0 - 1 - a.d0;
        tempvar range_check_ptr = range_check_ptr + 1;
    } else {
        assert [range_check_ptr] = p.d1 - 1 - a.d1;
        tempvar range_check_ptr = range_check_ptr + 1;
    }
} else {
    assert [range_check_ptr] = p.d2 - 1 - a.d2;
    tempvar range_check_ptr = range_check_ptr + 1;  
}
```